lib/ansible/modules/windows/win_domain_computer.ps1

```diff
@@ -0,0 +1,183 @@
+$diff_support = Get-AnsibleParam -obj $params -name "_ansible_diff" -default $false | ConvertTo-Bool
+
+# ------------------------------------------------------------------------------
+Function Get-DesiredState($params) {
+  $name = Get-AnsibleParam -obj $params -name "name" -failifempty $true -resultobj $result
+  $sam_account_name = Get-AnsibleParam -obj $params -name "sam_account_name" -default "$name$"
+  If (-not $sam_account_name.EndsWith("$")) {
+    Fail-Json $result "sam_account_name must end in $"
+  }
+  $enabled = Get-AnsibleParam -obj $params -name "enabled" -default $true | ConvertTo-Bool
+  $description = Get-AnsibleParam -obj $params -name "description" -default "" -resultobj $result
+  $state =(Get-AnsibleParam -obj $params -name "state" -ValidateSet "Present","Absent" -default "Present").ToLower()
+  If ($state -eq "present") {
+    $dns_hostname = Get-AnsibleParam -obj $params -name "dns_hostname" -failifempty $true -resultobj $result
+    $ou = Get-AnsibleParam -obj $params -name "ou" -failifempty $true -resultobj $result
+    $distinguished_name = "CN=$name,$ou"
+
+    $desired_state = @{
+      name = $name
+      sam_account_name = $sam_account_name
+      dns_hostname = $dns_hostname
+      ou = $ou
+      distinguished_name = $distinguished_name
+      description = $description
+      enabled = $enabled
+      state = $state
+    }
+  } Else {
+    $desired_state = @{
+      name = $name
+      state = $state
+    }
+  }
+
+  Return $desired_state
+}
+# ------------------------------------------------------------------------------
+Function Get-InitialState($desired_state) {
+  # Test computer exists
+  $computer = Try { Get-ADComputer `
+      -Identity $desired_state.name `
+      -Properties DistinguishedName,DNSHostName,Enabled,Name,SamAccountName,Description,ObjectClass
+    } Catch { $null }
+  If ($computer) {
+      $initial_state = @{
+      name = $computer.Name
+      sam_account_name = $computer.SamAccountName
+      dns_hostname = $computer.DNSHostName
+      # Get OU from regexp that removes all characters to the first ","
+      ou = $computer.DistinguishedName -creplace "^[^,]*,",""
+      distinguished_name = $computer.DistinguishedName
+      description = $computer.Description
+      enabled = $computer.Enabled
+      state = "present"
+    }
+  } Else {
+    $initial_state = @{
+      name = $desired_state.name
+      state = "absent"
+    }
+  }
+
+  return $initial_state
+}
+
+# ------------------------------------------------------------------------------
+Function Set-ConstructedState($initial_state, $desired_state) {
+  Try {
+    Set-ADComputer `
+      -Identity $desired_state.name `
+      -SamAccountName $desired_state.name `
+      -DNSHostName $desired_state.dns_hostname `
+      -Enabled $desired_state.enabled `
+      -Description $desired_state.description `
+      -WhatIf:$check_mode
+  } Catch { Fail-Json $result "Error in function Set-ConstructedState" }
```

Comment:
When creating error messages you need to keep in mind the audience is for the end user, making them descriptive of the action that failed and some more info would be what you want to achieve. In this case (and the other catches below) I would;

* Change the default error from saying there was an error around the parent function run but to what it was trying to do
* Add some more descriptive info such as the identity/name that was being modified
* Show the actual exception message at the end of the msg, this helps both the user and people trying to maintain the module

In the end it would look more like this
```
} catch {
    Fail-Json -obj $result -msg "Failed to set the AD object $($desired_state.name): $($_.Exception.Message)"`
}
```

Instead of saying `Set-ConstructedState` failed, I can immediately see

* What the action was that failed, trying to set an AD object
* The AD object that I was trying to set, `$desired_state.name` to give me some context and basic error checking to make sure there wasn't a typo
* If I am more technically inclined I can see the underlying powershell error and see if I might be able to fix it myself or if it is an environmental issue.

I would do this for try/catches you have in this module as well.